src/web/headers.rs

```diff
@@ -22,23 +43,57 @@ impl Header for CanonicalUrl {
     where
         E: Extend<HeaderValue>,
     {
-        let value: HeaderValue = format!(r#"<{}>; rel="canonical""#, self.0).parse().unwrap();
+        let value: HeaderValue = format!(r#"<{}>; rel="canonical""#, self.build_full_uri())
+            .parse()
+            .unwrap();
 
         values.extend(std::iter::once(value));
     }
 }
 
+impl Serialize for CanonicalUrl {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&self.build_full_uri().to_string())
```

Comment:
are you sure it's correct to percent-encode here? I worry the template library will also percent-encode and we'll end up double-encoding.

Reply:
we're not encoding all characters, but only the ones we want to. 

If someone would forget the `|safe` (for example in `/crate/builds.html`) when including the canonical URL, other characters would probably would be encoded, so the URL would be unuseable anyways. 

URL percent encoding is encoding different chars than HTML percent encoding. (though some would probably overlap). 

Moving the encoding to the template is tricky, since this would involve serializing a broken URL into the context, and probably taking a tera filter to re-parse & correctly encode it (and mark as safe). 

What feels like the better approach to you?

Reply:
ah ok, sounds good 👍 - I don't think it's important to have encoding happen in the Tera templates, I just didn't realize the URL percent-encoding was  different from the HTML escaping.